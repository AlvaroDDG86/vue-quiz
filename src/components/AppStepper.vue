<template>
  <div class="app-stepper">
    <div class="app-stepper__container">
      <div class="app-stepper__progress" ref="progress"></div>
      <div
        v-for="step in steps"
        :key="step"
        class="app-stepper__step"
        :class="{ 'app-stepper__step--active': current >= step }"
      >
        <span class="app-stepper__number">{{ step }}</span>
      </div>
    </div>
    <div class="app-stepper__actions">
      <AppButton @click="prevHandler" :disabled="current === 1 || disablePrev"
        ><fa icon="arrow-left"
      /></AppButton>
      <AppButton
        @click="nextHandler"
        :disabled="current === steps || disableNext"
        ><fa icon="arrow-right"
      /></AppButton>
    </div>
  </div>
</template>
<script lang="ts">
import { defineComponent, ref, toRefs, watch } from "vue";
import AppButton from "./AppButton.vue";
export default defineComponent({
  name: "AppStepper",
  components: { AppButton },
  props: {
    steps: {
      type: Number,
      required: true,
    },
    current: {
      type: Number,
      required: true,
    },
    disablePrev: {
      type: Boolean,
      default: false,
      require: false,
    },
    disableNext: {
      type: Boolean,
      default: false,
      require: false,
    },
  },
  setup(props, { emit }) {
    const progress = ref();
    const { current, steps } = toRefs(props);
    watch(current, (currentValue) => {
      progress.value.style.width = `${
        (100 / (steps.value - 1)) * (currentValue - 1)
      }%`;
    });
    const prevHandler = () => {
      emit("prev");
    };
    const nextHandler = () => {
      emit("next");
    };
    return {
      progress,
      nextHandler,
      prevHandler,
    };
  },
});
</script>
<style lang="postcss">
.app-stepper {
  @apply w-full;
}
.app-stepper__container {
  @apply flex justify-between relative mb-8;
}

.app-stepper::before {
  @apply bg-blue-700 absolute top-1/2 left-0 -translate-y-1/2 h-1 w-full z-10;
}

.app-stepper__progress {
  top: calc(50% - 2px);
  @apply bg-blue-700 absolute left-0 h-1 duration-300;
}

.app-stepper__step {
  @apply flex justify-center items-center w-2 h-2 rounded-lg bg-white font-bold text-gray-700 z-10 border-2 border-blue-700 text-xs;
  @apply md:w-8 md:h-8 md:text-sm md:rounded-md;
}

.app-stepper__number {
  @apply hidden;
  @apply md:inline;
}

.app-stepper__step--active {
  @apply bg-blue-700 text-white duration-300;
}

.app-stepper__actions {
  @apply w-full flex justify-evenly;
}
</style>
