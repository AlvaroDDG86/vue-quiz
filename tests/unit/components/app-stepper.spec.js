import { shallowMount } from "@vue/test-utils";
import AppStepper from "@/components/AppStepper";

describe("AppStepper", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(AppStepper, {
      propsData: {
        steps: 4,
        current: 1,
      },
      global: {
        stubs: {
          fa: true,
          AppButton: true,
        },
      },
    });
  });
  test("Should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("should display the steps", () => {
    expect(wrapper.find(".app-stepper__step").exists()).toBe(true);
  });
  test("should disabled prev when step is 1", async () => {
    const directions = wrapper.findAll("app-button-stub");
    const prev = directions.at(0);
    expect(prev.attributes().disabled).toBeTruthy();
    await wrapper.setProps({ current: 4 });
    const next = directions.at(1);
    expect(next.attributes().disabled).toBeTruthy();
  });
  test("should prev and next", () => {
    const directions = wrapper.findAll("app-button-stub");
    directions.at(0).trigger("click");
    directions.at(1).trigger("click");
    expect(wrapper.emitted("prev").length).toBe(1);
    expect(wrapper.emitted("next").length).toBe(1);
  });
});
