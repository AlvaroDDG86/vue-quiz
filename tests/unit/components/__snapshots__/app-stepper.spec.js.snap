// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppStepper Should match snapshot 1`] = `
<div class="app-stepper">
  <div class="app-stepper__container">
    <div class="app-stepper__progress"></div>
    <div class="app-stepper__step app-stepper__step--active"><span class="app-stepper__number">1</span></div>
    <div class="app-stepper__step"><span class="app-stepper__number">2</span></div>
    <div class="app-stepper__step"><span class="app-stepper__number">3</span></div>
    <div class="app-stepper__step"><span class="app-stepper__number">4</span></div>
  </div>
  <div class="app-stepper__actions">
    <app-button-stub disabled="true"></app-button-stub>
    <app-button-stub disabled="false"></app-button-stub>
  </div>
</div>
`;
